vm/src/vm/vm_core.rs

```diff
@@ -316,13 +316,14 @@ impl VirtualMachine {
     fn deduce_dst(
         &self,
         instruction: &Instruction,
-        res: Option<&MaybeRelocatable>,
-    ) -> Option<MaybeRelocatable> {
-        match instruction.opcode {
-            Opcode::AssertEq => res.cloned(),
-            Opcode::Call => Some(self.get_fp().into()),
-            _ => None,
-        }
+        res: &Option<MaybeRelocatable>,
+    ) -> Result<MaybeRelocatable, VirtualMachineError> {
+        let dst = match instruction.opcode {
+            Opcode::AssertEq if res.is_some() => res.clone().unwrap(),
+            Opcode::Call => MaybeRelocatable::from(self.run_context.get_fp()),
+            _ => return Err(VirtualMachineError::NoDst),
+        };
```

Comment:
Maybe we could avoid the `unwrap` if we match both arguments at the same time:
```suggestion
        let dst = match (instruction.opcode, res) {
            (Opcode::AssertEq, Some(res)) => res.clone(),
            (Opcode::Call, _) => MaybeRelocatable::from(self.run_context.get_fp()),
            _ => return Err(VirtualMachineError::NoDst),
        };
```